### PR TITLE
Added VicReg - Variance-Invariance-Covariance Regularization for SSL

### PR DESCRIPTION
**Pull request by dewball345**
This is a draft PR - updated .py files will be added

### CODE REVIEW

**Comment by fchollet** [comment]
Any review you need on my side? Or from the TF-Similarity authors?

**Comment by dewball345** [comment]
Hi @fchollet

I'd want your feedback as well as comments from @ebursztein or @owenvallis though I still need to upload the notebook(hence why it's a draft).

**Comment by fchollet** [comment]
The changes look good to me.  @ebursztein & @owenvallis -- please chime in!

**Comment by dewball345** [comment]
> The changes look good to me. @ebursztein & @owenvallis -- please chime in!

Hi @fchollet, I did not include the actual file yet. Just have to do some grammar tweaks before I upload the notebook

**Comment by dewball345** [comment]
@owenvallis I've added the VicReg example. BTW - have used a custom fork due to some changes in augmenters. I've made a PR at the tf similarity repo so if that gets merged to either the dev or master branch I'll change the code to use it instead.

**Comment by dewball345** [comment]
@owenvallis @ebursztein could you check the VicReg code to see if everything looks good?

File: examples/vision/ipynb/barlow_twins.ipynb
Comment (owenvallis) [inline examples/vision/ipynb/barlow_twins.ipynb:192]: Does this need the double !! for the shell command?

**Review by owenvallis — APPROVED** [review]

**Comment by owenvallis** [comment]
Small note on the pip install commands, but looks good to me otherwise. Thanks again for adding the update.

**Comment by dewball345** [comment]
Okay, I've made some changes - removed Barlow twins commits because those are unrelated to VicReg. Also, the TensorFlow similarity master branch is now used instead of the custom fork. 

Let me know if everything looks okay before I create an ipynb notebook

**Review by owenvallis — APPROVED** [review]

**Comment by dewball345** [comment]
> Small note on the pip install commands, but looks good to me otherwise. Thanks again for adding the update.

Hey owen, have also added my VicReg files - LMK if they are okay and we can merge.

File: examples/vision/md/vicreg.md
Comment (owenvallis) [inline examples/vision/md/vicreg.md:105]: I think we likely want to remove these repeated output lines.

**Review by owenvallis — CHANGES_REQUESTED** [review]
Looks good. I left a note that you might want to clean the output of the md and ipynb files as it looks like some code blocks printed a large number of lines.

File: examples/vision/md/vicreg.md
Comment (dewball345) [reply examples/vision/md/vicreg.md]: @owenvallis  I've finished this

Comment (owenvallis) [reply examples/vision/md/vicreg.md]: thanks, this looks better, although I would still probably remove most of the loss outputs. Maybe just keep just the first 2 and the last 2. Up to you though. You can always update the output if you change your mind.

@fchollet, what's the recommendation regarding printing loss outputs in the md and ipynb files?

**Review by owenvallis — APPROVED** [review]

File: examples/vision/md/vicreg.md
Comment (dewball345) [reply examples/vision/md/vicreg.md]: @owenvallis From what I found in the Keras io examples(including what I did in my previous examples), the loss outputs are kept as they are? Probably would be good to ask @fchollet but I'm open to including the first two and last two for readability purposes

Comment (dewball345) [reply examples/vision/md/vicreg.md]: @fchollet @owenvallis  Any update on this?

**Review by owenvallis — APPROVED** [review]

**Review by owenvallis — APPROVED** [review]

File: examples/vision/md/vicreg.md
Comment (owenvallis) [reply examples/vision/md/vicreg.md]: Added approval for the changes, but looks like you also need approval from someone with write access to the repo.

**Comment by dewball345** [comment]
@fchollet Could you approve/give feedback on this PR - LMK if there are any issues with it.

**Comment by dewball345** [comment]
Closing and will bump this up

**Pull request closed without merging**
